include/common/core/allocator.hpp

```diff
@@ -0,0 +1,84 @@
+#include <algorithm>
+#include <array>
+#include <variant>
+
+#include "common/core/logging.h"
+
+/*
+ * A generic pool allocator class that can be used to statically allocate memory
+ * for typically dynamic objects (such as maps).
+ *
+ * Example Use-Case:
+ *
+ *
+ */
+
+template <typename Element>
+struct AllocatorElement {
+    // The address of element member must be the same as the address of an
+    // AllocatorElement.
+    // Thus, element must be first member of this struct. And there can be no
+    // member functions and no private variables.
+    // See https://en.cppreference.com/w/cpp/types/is_standard_layout
+    Element element{};
+    bool used = false;
+};
+
+// should have type of object and max number of.
+// should have backing be an array and/or pointer to array. Do not use malloc!
+// only support allocating memory for one type of object at a time.
+
+// Use concept to check that individual size
+template <typename Element, std::size_t MaxElements>
+class PoolAllocator {
+  public:
+    using value_type = Element;
+    using size_type = std::size_t;
+    using difference_type = ptrdiff_t;
+    using pointer = Element*;
+    using const_pointer = const Element*;
+    using reference = Element&;
+    using const_reference = const Element&;
+
+    PoolAllocator() {}
+    auto operator=(PoolAllocator&) -> PoolAllocator& = delete;
+    auto operator=(PoolAllocator&&) -> PoolAllocator&& = delete;
+    PoolAllocator(const PoolAllocator&) = default;
+    PoolAllocator(PoolAllocator&&) = delete;
+    ~PoolAllocator() = default;
+
+    void deallocate(pointer loc, size_t) {
+        // Find index into backing array using the fact that loc is the first
+        // item in the AllocatorElement struct and so has the same address.
+        auto index = (reinterpret_cast<AllocatorElement<Element>*>(loc) -
+                      backing.begin()) /
+                     sizeof(AllocatorElement<Element>);
+        LOG("Deallocating %d", index);
+
+        if (index < MaxElements) {
+            backing[index].used = false;
+        }
+    }
+
+    pointer allocate(size_t, const void* = 0) {
```

Comment:
Since there's no implementation here for `max_size()`, isn't there a chance that `allocate()` will be invoked to allocate more than one item at a time?

Reply:
Hm I don't think allocate would be attempting to allocate more than one item, but we should definitely be checking that the size given doesn't exceed the max size of the element. Will probably end up doing something like [seth suggested](https://github.com/Opentrons/ot3-firmware/pull/376#discussion_r886798397).

Reply:
`vector`, `deque`, and some other container types allocate arrays instead of individual items, which means `allocate` would attempt to allocate more than one item.

I think that the type of your `backing` array will have to change if you want to support block allocation with Seth's suggestion. If you allocate out a block of multiple contiguous elements in the current backing array, you're giving the caller free rein to write over the `AllocatorElement::used` members.

Reply:
Oh that's a very good point, you're right.